phones/tests/models_tests.py

```diff
@@ -315,37 +345,191 @@ def test_create_relaynumber_already_registered_with_service(
+    )
+    mock_services.assert_called_once_with(settings.TWILIO_MESSAGING_SERVICE_SID[0])
+    mock_messaging_number_create.assert_called_once_with(
+        phone_number_sid=twilio_incoming_number_sid
+    )
     mock_messages_create.assert_called_once()
-    mock_messages_create.reset_mock()
+    assert caplog.messages == ["twilio_messaging_service"]
+    assert caplog.records[0].code == 21710
 
-    relay_number = "+19998887777"
-    relay_number_obj = RelayNumber.objects.create(user=phone_user, number=relay_number)
 
-    mock_number_create.assert_called_once()
-    call_kwargs = mock_number_create.call_args.kwargs
-    assert call_kwargs["phone_number"] == relay_number
-    assert call_kwargs["sms_application_sid"] == settings.TWILIO_SMS_APPLICATION_SID
-    assert call_kwargs["voice_application_sid"] == settings.TWILIO_SMS_APPLICATION_SID
+def test_create_relaynumber_full_service(
+    phone_user, real_phone_us, mock_twilio_client, test_settings, caplog
+):
+    """If the Twilio Messaging Service pool is full, an exception is raised."""
+    twilio_incoming_number_sid = "PNXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+    twilio_service_sid = test_settings.TWILIO_MESSAGING_SERVICE_SID[0]
+    mock_messages_create = mock_twilio_client.messages.create
+    mock_number_create = mock_twilio_client.incoming_phone_numbers.create
+    mock_number_create.return_value = SimpleNamespace(sid=twilio_incoming_number_sid)
+    mock_services = mock_twilio_client.messaging.v1.services
+    mock_messaging_number_create = mock_services.return_value.phone_numbers.create
 
-    mock_services.assert_called_once()
-    call_args = mock_services.call_args
-    assert call_args[0][0] == settings.TWILIO_MESSAGING_SERVICE_SID
-    mock_messaging_number_create.assert_called_once()
-    call_kwargs = mock_messaging_number_create.call_args.kwargs
-    assert call_kwargs["phone_number_sid"] == twilio_incoming_number_sid
+    # Twilio responds that the pool is full
+    mock_messaging_number_create.side_effect = TwilioRestException(
+        uri=f"/Services/{twilio_service_sid}/PhoneNumbers",
+        msg=("Unable to create record: Number Pool size limit reached"),
+        method="POST",
+        status=412,
+        code=21714,
+    )
+
+    # "Pool full" exception is raised
+    with pytest.raises(Exception) as exc_info:
+        RelayNumber.objects.create(user=phone_user, number="+19998887777")
+    assert (
+        str(exc_info.value) == "All services in TWILIO_MESSAGING_SERVICE_SID are full"
+    )
+
+    mock_messaging_number_create.assert_called_once_with(
+        phone_number_sid=twilio_incoming_number_sid
+    )
+    mock_messages_create.assert_not_called()
+    assert caplog.messages == ["twilio_messaging_service"]
+    assert caplog.records[0].code == 21714
+
+
+def test_create_relaynumber_no_service(
+    phone_user, real_phone_us, mock_twilio_client, test_settings, caplog
+):
+    """If no Twilio Messaging Service IDs are defined, registration is skipped."""
```

Comment:
comment (non-blocking): Interesting. This means the numbers will not have a service ID, so they won't be in a 10DLC campaign. Which is probaby fine for dev instances

Reply:
That was what I thought would be the most common case. However, it does add something we need to be careful about in the deployments (which we're about to update with multiple SIDs anyway)

I see a few options:

1. Skip registration when no service ID is given (this code)
2. Fail if no service ID is given (the phone number is already reserved, however) - requires updating other tests to pass
3. Add an invalid service ID as the default (tests pass, but deployments / local dev fail until a real number is entered)
4. Add a valid service ID as the default (tests pass, real use may fail if service is full, but after allocating number)

I think "no service ID = no registration" might be the best quick option.

Reply:
Yeah, that makes sense.